crypto/fipsmodule/ec/p384.c

```diff
@@ -0,0 +1,452 @@
+static void fiat_p384_point_add(fiat_p384_felem x3, fiat_p384_felem y3,
+                                fiat_p384_felem z3, const fiat_p384_felem x1,
+                                const fiat_p384_felem y1,
+                                const fiat_p384_felem z1, const int mixed,
+                                const fiat_p384_felem x2,
+                                const fiat_p384_felem y2,
+                                const fiat_p384_felem z2) {
+  fiat_p384_felem x_out, y_out, z_out;
+  fiat_p384_limb_t z1nz = fiat_p384_nz(z1);
+  fiat_p384_limb_t z2nz = fiat_p384_nz(z2);
+
+  // z1z1 = z1z1 = z1**2
+  fiat_p384_felem z1z1;
+  fiat_p384_square(z1z1, z1);
+
+  fiat_p384_felem u1, s1, two_z1z2;
+  if (!mixed) {
+    // z2z2 = z2**2
+    fiat_p384_felem z2z2;
+    fiat_p384_square(z2z2, z2);
+
+    // u1 = x1*z2z2
+    fiat_p384_mul(u1, x1, z2z2);
+
+    // two_z1z2 = (z1 + z2)**2 - (z1z1 + z2z2) = 2z1z2
+    fiat_p384_add(two_z1z2, z1, z2);
+    fiat_p384_square(two_z1z2, two_z1z2);
+    fiat_p384_sub(two_z1z2, two_z1z2, z1z1);
+    fiat_p384_sub(two_z1z2, two_z1z2, z2z2);
+
+    // s1 = y1 * z2**3
+    fiat_p384_mul(s1, z2, z2z2);
+    fiat_p384_mul(s1, s1, y1);
+  } else {
+    // We'll assume z2 = 1 (special case z2 = 0 is handled later).
+
+    // u1 = x1*z2z2
+    fiat_p384_copy(u1, x1);
+    // two_z1z2 = 2z1z2
+    fiat_p384_add(two_z1z2, z1, z1);
+    // s1 = y1 * z2**3
+    fiat_p384_copy(s1, y1);
+  }
+
+  // u2 = x2*z1z1
+  fiat_p384_felem u2;
+  fiat_p384_mul(u2, x2, z1z1);
+
+  // h = u2 - u1
+  fiat_p384_felem h;
+  fiat_p384_sub(h, u2, u1);
+
+  fiat_p384_limb_t xneq = fiat_p384_nz(h);
+
+  // z_out = two_z1z2 * h
+  fiat_p384_mul(z_out, h, two_z1z2);
+
+  // z1z1z1 = z1 * z1z1
+  fiat_p384_felem z1z1z1;
+  fiat_p384_mul(z1z1z1, z1, z1z1);
+
+  // s2 = y2 * z1**3
+  fiat_p384_felem s2;
+  fiat_p384_mul(s2, y2, z1z1z1);
+
+  // r = (s2 - s1)*2
+  fiat_p384_felem r;
+  fiat_p384_sub(r, s2, s1);
+  fiat_p384_add(r, r, r);
+
+  fiat_p384_limb_t yneq = fiat_p384_nz(r);
+
+  // This case will never occur in the constant-time |ec_GFp_mont_mul|.
+  fiat_p384_limb_t is_nontrivial_double = constant_time_is_zero_w(xneq | yneq) &
+                                          ~constant_time_is_zero_w(z1nz) &
+                                          ~constant_time_is_zero_w(z2nz);
+  if (is_nontrivial_double) {
+    fiat_p384_point_double(x3, y3, z3, x1, y1, z1);
+    return;
+  }
+
+  // I = (2h)**2
+  fiat_p384_felem i;
+  fiat_p384_add(i, h, h);
+  fiat_p384_square(i, i);
+
+  // J = h * I
+  fiat_p384_felem j;
+  fiat_p384_mul(j, h, i);
+
+  // V = U1 * I
+  fiat_p384_felem v;
+  fiat_p384_mul(v, u1, i);
+
+  // x_out = r**2 - J - 2V
+  fiat_p384_square(x_out, r);
+  fiat_p384_sub(x_out, x_out, j);
+  fiat_p384_sub(x_out, x_out, v);
+  fiat_p384_sub(x_out, x_out, v);
+
+  // y_out = r(V-x_out) - 2 * s1 * J
+  fiat_p384_sub(y_out, v, x_out);
+  fiat_p384_mul(y_out, y_out, r);
+  fiat_p384_felem s1j;
```

Comment:
You're right, but I think I'll leave it as is, since it's not increasing the stack usage in comparison to other functions, especially those that will do scalar multiplication.